pantheon-bundle/src/main/java/com/redhat/pantheon/asciidoctor/extension/XrefPreprocessor.java

```diff
@@ -0,0 +1,122 @@
+package com.redhat.pantheon.asciidoctor.extension;
+
+import com.redhat.pantheon.helper.PantheonConstants;
+import com.redhat.pantheon.model.assembly.TableOfContents;
+import com.redhat.pantheon.model.document.Document;
+import com.redhat.pantheon.model.document.DocumentLocale;
+import com.redhat.pantheon.model.document.DocumentVariant;
+import com.redhat.pantheon.model.module.Module;
+import org.apache.sling.api.resource.Resource;
+import org.asciidoctor.extension.Preprocessor;
+import org.asciidoctor.extension.PreprocessorReader;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class XrefPreprocessor extends Preprocessor {
+
+    private static final Logger log = LoggerFactory.getLogger(XrefPreprocessor.class);
+
+    private static final Pattern XREF_PATTERN = Pattern.compile("xref:(?<filepath>\\S*?)(?:#(?<anchor>\\S*))?\\[(?<label>.*?)\\]");
+    private static final Pattern TRIANGLE_PATTERN = Pattern.compile("<<(?<filepath>\\S*?)(?:#(?<anchor>\\S*))?,(?<label>.*?)>>");
+
+    private DocumentVariant documentVariant;
+    private TableOfContents toc;
+    private Set<String> modulePaths = new HashSet<>();
+
+    /**
+     * @param documentVariant
+     * @param tableOfContents Doesn't necessarily need to be populated when this constructor is called.
+     */
+    public XrefPreprocessor(DocumentVariant documentVariant, TableOfContents tableOfContents) {
+        this.documentVariant = documentVariant;
+        this.toc = tableOfContents;
+    }
+
+    @Override
+    public void process(org.asciidoctor.ast.Document adocDocument, @NotNull PreprocessorReader reader) {
+        reader.restoreLines(preprocess(reader.readLines()));
+    }
+
+    private String processLineWithPattern(String line, Pattern pattern) {
+        Matcher matcher = pattern.matcher(line);
+        StringBuffer sb = new StringBuffer();
+        while (matcher.find()) {
+            String originalTarget = matcher.group("filepath");
+            // Assume it's a relative path to a file in the same repo for now
+            Resource desiredTarget = documentVariant.getResourceResolver().getResource(documentVariant.getParentLocale().getParent().getParent().getPath() + "/" + originalTarget);
```

Comment:
if would be super if we can add some line wrap here to make it more readable :-)

Reply:
Okie dokie, I broke up the line a bit with the latest commit.